### ISSUE TITLE
Git notes being rejected due to missing GPG signature

### ISSUE THREAD

**Issue by maryiam** [semantic-release/semantic-release#1479]
## Current behavior

After upgrading to semantic-release `^16.0.0`, I'm not able to release my packages through Gitlab CI.
The job fails and I see the following error in the logs:

> [semantic-release] › ✖  An error occurred while running semantic-release: Error: Command failed with exit code 1: git push git@xxx/yyy.git refs/notes/semantic-release.  
Remote: GitLab: Commit must be signed with a GPG key        

## Environment

- **semantic-release** version: `v16.0.4`
- CI environment: `Gitlab`
- Plugins used: `[
      "@semantic-release/commit-analyzer",
      "@semantic-release/release-notes-generator",
      "@semantic-release/changelog",
      "@semantic-release/npm",
      "@semantic-release/git"
    ]`
- The option `Reject unsigned commits` is checked in my Gitlab project settings
- The ordinary release commit with the format `chore(release): x.x.x [skip ci] ` that is made by semantic-release is committed, signed and pushed. Also, the tag is created and pushed. But the npm publish is not happening (semantic-release fails with pushing git notes before getting to the publish step).

I'm not able to understand why is this happening? Do we lose the GPG signature once we add a git note to the commit? I don't want to remove the option `Reject unsigned commit` form my settings. Can I choose to not use the git notes with semantic-release? 

Thanks.

**Comment by pvdlg** [comment]
That seems related to #1446.
It seems GitLab allow to set some checks on commits but is unable to distinguish  a note from a commit.
I would recommend to open an issue with GitLab.

> I'm not able to understand why is this happening? Do we lose the GPG signature once we add a git note to the commit?
I assume GitLab wrongly interpret notes as commits.

> Can I choose to not use the git notes with semantic-release?
No, semantic-release use those to maintain a state (what was published where) in between runs.

**Comment by greg0ire** [comment]
@pvdlg hi! Sorry for digging this up. The implication of saying that this is an issue with Gitlab is that there would be no security implications of allowing to push unsigned git notes. I'm not sure that is true. For instance, what if an attacker were to push a git note with a malicious payload, and then `semantic-release` would read that note?

IMO it's important that the store where semrel stores its metadata be secure and trusted.

If that's not easy to do, then please reconsider #1646 . Right now, when I look at my repository, it seems like I will only every have `{"channels":[null]}` stored in there.

The default value seems to be `[null]` as well: https://github.com/semantic-release/semantic-release/blob/971a5e0d16f1a32e117e9ce382a1618c8256d0d9/lib/branches/get-tags.js#L21-L23

That makes me thing I could do without the notes, and that is confirmed by what I read here: https://dev.to/antongolub/you-don-t-need-semantic-release-sometimes-3k6k

Should the notes part of this code block be skipped when it's going to store `[null]`? https://github.com/semantic-release/semantic-release/blob/971a5e0d16f1a32e117e9ce382a1618c8256d0d9/index.js#L115-L117

**Comment by Langstra** [comment]
This is an issue with Gitlab, also see this issue. I has been posted before in other issue, but just to make sure it can be found easily. https://gitlab.com/gitlab-org/gitlab/-/issues/207807

**Comment by greg0ire** [comment]
@Langstra the issue you linked to is about a commit being rejected due to its commit message. Here it's being rejected due to a lack of GPG signature, which is something else entirely. Maybe the issue is with Gitlab, but if it is, it might not have been reported yet.

**Comment by Langstra** [comment]
Afaik it is impossible to sign git notes. https://git-scm.com/docs/git-notes There is no mentioning of the possibility in the documentation. While as, if you have enabled `Reject unsigned commits` in Gitlab, Gitlab also rejects unsigned notes. That is what issues 207807 is about and why I thought it was relevant to mention here.

**Comment by greg0ire** [comment]
You're right, there is indeed a thread that mentions it, I should have read more thoroughly: https://gitlab.com/gitlab-org/gitlab/-/issues/207807#note_569320409

**Comment by juancarlosjr97** [comment]
Any updates on this?

**Comment by juancarlosjr97** [comment]
The change to be introduced to semantic release is minimum and could only apply to GitLab. Due to Gitlab taking a lot of time to fix. Can we introduce that change ourselves? So many people cannot use Semantic Release on Gitlab due to not working when signed commits are enforced :(

**Comment by LuqWright** [comment]
So Gitlab is saying that this issue is resolved, but I can't seem to get it to work. The GL issue is closed, but this one is still opened. Has anyone ever been able to get it working?

Link to GL issue here -- https://gitlab.com/gitlab-org/gitlab/-/issues/207807#note_569320409

I'm wondering if Gitlab actually _did_ resolve this, but then semantic-release made a change that broke it?

**Comment by travi** [comment]
@LuqWright this issue is specific to the failure to push _notes_, which sounds like it is fixed on the gitlab side, based on that GL issue. are you sure the failure you are seeing is related to pushing _notes_ and not _commits_? what is the specific error message you are seeing?

> but then semantic-release made a change that broke it?

this is very unlikely. signing of git notes is not possible and has never been possible. our behavior related to this has not changed

**Comment by LuqWright** [comment]
@travi Here is the exact  error

```
[
    undefined,
    '',
    'remote: \n' +
      'remote: Error: Commit a4251f705a8b49ad7337273c88e21a353c42c1e2 is not signed.        \n' +
      'remote: See documentation at https://[PROJECT_URL]/welcome/guides/-/blob/main/commit-signing.md        \n' +
      'remote: \n' +
      'To https://[PROJECT_URL]/project.git\n' +
      ' ! [remote rejected]     refs/notes/semantic-release-v1.0.26 -> refs/notes/semantic-release-v1.0.26 (pre-receive hook declined)\n' +
      "error: failed to push some refs to 'https://[PROJECT_URL]/project.git'"
  ]
```

Commit-signing is enforced across the entire GitLab instance, and since the error message mentions notes and signing, I suspect it may be related to that.

**Comment by travi** [comment]
i would agree that appears to be related to notes. i'm not a gitlab user, but my understanding is that this is unrelated to semantic-release and entirely a problem on the gitlab side. the only thing that comes to mind is ensuring your gitlab instance is on a version that includes the fix mentioned in the gitlab ticket

**Comment by LuqWright** [comment]
@travi Thank you for the assist, that got me digging deeper into the code and I believe I may have found it. Looks like Gitlab's fix specifically excludes `refs/notes/commits` instead of all notes `refs/notes/*`

The comment saying that they were able to get it working -- the only thing I can think of is if they were overriding the `GIT_NOTE_REF` variable to `commits` before the change of `refs/notes/${GIT_NOTE_REF}` to `refs/notes/${GIT_NOTE_REF}-${ref}` in [lib/git.js](https://github.com/semantic-release/semantic-release/commit/020ea7d7d5aa6c123fe4c97754b6b1b36c0a8e2c)

@juancarlosjr97 Are you able to confirm if that's how you got it working previously? I know it was years ago haha..I left a comment on the GL issue though -- this is definitely something on their end. Thank you again

**Comment by travi** [comment]
> Looks like Gitlab's fix specifically excludes `refs/notes/commits` instead of all notes `refs/notes/*`

i appreciate the follow up with additional information. if you find more detail, please continue to share here so others can find it

> this is definitely something on their end

because of this, i think it is appropriate to close this issue. there is nothing that the semantic-release project can take action on related to this. please feel free to continue adding details here beyond the actions the project can take, though.
